### PR TITLE
nixos/snipe-it: Add missing upload directories

### LINKED ISSUE #187179: Snipe-IT: barcode generation fails due to missing directory
### Describe the bug
Snipe-IT needs the directory public/uploads/barcodes/ in order to generate barcodes and QR codes for asset labels. This directory should be created during installation, but isn't. Creating this directory manually fixes the problem.

### Steps To Reproduce
1. Setup a Snipe-IT instance.
2. Add an asset.
3. Generate an asset label. 

### Expected behavior
The directory (public/uploads/barcodes/) should exist and have the same content as in the Snipe-IT GitHub repository.

### Notify maintainers

@yayayayaka 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.88, NixOS, 22.05 (Quokka), 22.05.2473.cfabaa15e98`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(root): `"nixos-22.05-small"`
 - channels(helmholtzschule): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by yayayayaka**
###### Description of changes

- Adds additional systemd-tmpfiles rules to ensure necessary upload folders exist.

Resolves #187179
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/snipe-it.nix [hunk 1]
```diff
@@ -454,25 +454,43 @@ in {
 
           # migrate db
           ${pkgs.php}/bin/php artisan migrate --force
+
+          # A placeholder file for invalid barcodes
+          invalid_barcode_location="${cfg.dataDir}/public/uploads/barcodes/invalid_barcode.gif"
+          [ ! -f "$invalid_barcode_location" ] \
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/snipe-it.nix:460]:
```suggestion
          [ ! -e "$invalid_barcode_location" ] \
```
Just to be safe to not overwrite symlinks

File: pkgs/servers/web-apps/snipe-it/default.nix [hunk 2]
```diff
@@ -7,12 +7,19 @@ let
     noDev = true; # Disable development dependencies
   }).overrideAttrs (attrs : {
     installPhase = attrs.installPhase + ''
+      # Before symlinking the following directories, copy the invalid_barcode.gif
+      # to a different location. The `snipe-it-setup` oneshot service will then
+      # copy the file back during bootstrap.
+      cp $out/public/uploads/barcodes/invalid_barcode.gif $out/
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/snipe-it/default.nix:13]:
```suggestion
      cp $out/public/uploads/barcodes/invalid_barcode.gif{,.backup}
```
That would be a bit nicer than just dumping it into $out

Comment (yayayayaka) [reply pkgs/servers/web-apps/snipe-it/default.nix:13]: We symlink the public/ directory with the dataDir so this likely won't work.

Comment (SuperSandro2000) [reply pkgs/servers/web-apps/snipe-it/default.nix:13]: Then maybe some other directory except plain $out? $out/share/snipe-it/invalid_barcode.gif

Comment (yayayayaka) [reply pkgs/servers/web-apps/snipe-it/default.nix:13]: Alright, let's do it that way.

**Pull request merged**